torch/csrc/autograd/record_function.h

```diff
@@ -103,6 +103,33 @@ struct TORCH_API RecordFunction {
 
   void end();
 
+  // Saves the thread_id that this RecordFunction was created with. This is
+  // needed so that we can access Events created by the original thread in a
+  // different thread, since they are thread-local. An example use case is
+  // calling RecordFunction::end() in a different thread.
+  inline void setThreadId(uint16_t threadId) const {
+    threadId_ = threadId;
+  }
+
+  // A flag controlling whether we should use the thread set by setThreadId() or
+  //  not in callback-specific logic.
+  inline bool overrideThreadId() const {
+    return overrideThreadId_;
+  }
+
+  // Sets a flag indicating whether we should use the thread set by
+  // wsetThreadId() or not in callback-specific logic.
+  inline void setOverrideThreadId(bool shouldOverride=true) const {
+    overrideThreadId_ = shouldOverride;
+  }
+
+  // Retrieves the thread_id that this RecordFunction was created with. Useful
+  // if we need to access Events created by the original thread in a different
+  // thread.
+  inline uint16_t getThreadId() const {
+    return threadId_;
```

Comment:
With the threadId_ changes that we discussed offline, make sure you add a comment explaining that the threadId_ is only set in cases where the end function is called from a different thread.

Reply:
I think we decided not to do that since we weren't sure if every RPC starting from python layer eventually calls into `sendMessageWithAutograd`. Although upon further inspection it looks like we can do this. Although a downside is we'd have to pass the shared_ptr another layer down,.

Reply:
Added.